### DISCUSSION TITLE
Getting a full list of steps?

### DISCUSSION THREAD

**Discussion by EGTomKies** [getgauge/gauge#2423] [General]
I know it loads in VSC Gauge will pre-register steps.  

Is there any way to utilize that to get a list of steps outputted to a file?  I'm trying to make a list for my user's benefit.  

Going line by line over the file and pulling out the steps in lines starting with `@step` is easy enough, but I'm having issues getting at the ones inside an array.  

Before I reinvent the wheel, I figured I'd ask if there's a CLI tool I may be overlooking like the one that can out put a full list of specs.

**Comment by sriv** [comment]
there is a plugin for gauge called [`spectacle`](https://github.com/getgauge/spectacle) that generates a searchable html with list of steps/scenarios -  https://github.com/getgauge/spectacle

this is the closest I can think of. There is also `gauge list` command, but that does not support steps

**Reply by EGTomKies** [reply]
I've seen that one, it's not quite what I need, but definitely occasionally useful

**Reply by sriv** [reply]
got it. Do you have any example of what sort of output you need? 

I guess there isn't something out of box, but I can try to see if this can be "hacked out". (Hint: we get a list of steps in the IDE for autocompletion etc. I am thinking of a util that can either use the LSP (like we do [in tests](https://github.com/getgauge/gauge-lsp-tests/blob/master/tests/codecomplete.js#L19)) or the protobuf based API and unmarshall the response.

**Comment by EGTomKies** [comment]
Sure, I appreciate the offer.    

Let's say I have the following : 

```
@step("Click the Send Mail button")
def click_send_mail_button():
   # some code here
   
@step([
     "Click Submit",
     "Click the Submit Button",
     "Click the element with accessibility id: Submit"
])
def click_submit():
    #  some more code here  
```

My goal is, to sweep the step_impl directory to get a list of every step available: 

* Click the Send Mail button
* Click Submit
* Click the Submit Button
* Click the element with accessibility id: Submit

These wind up going into a user guide we have for our test developers

The sticking point is getting at the ones with multiple aliases.    I even tried ChatGPT for a regex to help, but so far that's been a dead end, I think it's hung up on the line breaks.  

```
# Specify the input file path
$inputFilePath = "C:\myproject\step_impl\shared_impl\steps.py"

# Specify the output file path
Set-Location C:\githubfiles\ecp\test_automation\step_impl\shared_impl\scripts
$outputFilePath = "output_file.txt"

# Define a regular expression pattern to match the quoted strings within "@step([...])"
$pattern = '@step\(\[\s*"([^"]+)"\s*,\s*"([^"]+)"\s*\]\)'

# Initialize an empty array to store the matched strings
$matches = @()

# Read the content of the input file line by line
Get-Content -Path $inputFilePath | ForEach-Object {
    # Use the Select-String cmdlet to find matches in the current line
    $lineMatches = $_ | Select-String -Pattern $pattern | ForEach-Object { $_.Matches.Groups[1].Value, $_.Matches.Groups[2].Value }
    # Add the matches from the current line to the array
    $matches += $lineMatches
}

# Write all matched strings to the output file
$matches | Out-File -FilePath $outputFilePath

# Display the matched strings (optional)
$matches
```

**Comment by sriv** [comment]
I made some changes to support `gauge list --steps` on my local and I can see this output for this project : https://github.com/getgauge/gauge-lsp-tests

```
~/projects/gauge-lsp-tests$ ../gauge/bin/linux_amd64/gauge list --steps
[Steps]
Start LSP and initialize - This should be the first request from the client to the server
cache gauge init template if not already present
copy data - env, specifications and implementation folders of required data from <dataProjectPath>
copy project template from cache
create temporary directory
execute gauge language runner pre-requisite
invoke shutDown and exit of LSP
open project and verify diagnostics <diagnostics>
remove the env, specs and impl folders copied from the template
remove the temporary directory
simulate user changing content on file system <relativeFilePath> to <contentFile> and saving on disk
textDocument/didChange event to change content from <relativeFilePath> to <contentFile>
textDocument/didOpen <file> with content <content>
textDocument/didOpen <fileName>
textDocument/didOpen details from <jsonDetails>
textDocument/didSave - The document save notification is sent from the client to the server when the document <relativeFilePath> was saved in the client.
verify gauge runner with call gauge/getRunnerLanguage
wait for "0.1" seconds
```

Is this the sort of output you were expecting? I can try to push the code to a branch if that helps

**Reply by EGTomKies** [reply]
That sort of basic output would be fine, I can copy it into something else to clean it up for presentation easy enough.  That would be awesome, thank you!    

We usually only install from the releases rather than building our own, would it make it out to there?

**Reply by sriv** [reply]
If you have a suggestion for an alternative format, please do share.

I plan to raise a PR and get this merged, once I have cleaned the code up and tested. What would be helpful is if you can verify if the change yields the desired list. I will drop a note once the PR is raised.

**Reply by EGTomKies** [reply]
A flat list is perfect, thank you, much appreciated

**Reply by sriv** [reply]
I realized that I was only fetching the list of steps from the spec files earlier, and you were looking to get the implementations as well (since you mentioned parsing aliases etc). Now I have something working, but there is a catch.

Lets say a step is defined in the implementation (java for this example) as such:

```java
    @Step({"Vowels in English language are <vowelString>.", "Vowels are <vowelString>."})
    public void setLanguageVowels(String vowelString) {
        vowels = new HashSet<>();
        for (char ch : vowelString.toCharArray()) {
            vowels.add(ch);
        }
    }
```

And is being used in the specification like this:

```markdown
# Specification Heading

This is an executable specification file. This file follows markdown syntax.
Every heading in this file denotes a scenario. Every bulleted point denotes a step.

To execute this specification, run
	gauge run specs


* Vowels in English language are "aeiou".

```

Now when I do `gauge list --steps` I see a response like this:

```console
Vowels are <vowelString>.
Vowels in English language are "aeiou".
Vowels in English language are <vowelString>.
```

The question is that the template used to define a step and use it are the same but the parameters passed differ.

The easiest way is to ignore the parameters and print the step template in the form of `Vowels in English language are {}.`

What do you feel about this? Does this help or hurt your scenario?

**Reply by EGTomKies** [reply]
Thank you for the update!  

We don't need the actual usage implementations, i.e. `Vowels in English language are "aeiou"`.

We'd only need the step definitions: 
```
Vowels are <vowelString>.
Vowels in English language are <vowelString>.
```

We have multiple departments/teams using the testing framework, and the goal is to get them a list of available steps so they know what's available before trying to create their own or filing a ticket with my group to create the step for them.   This way they have a catalog of steps to choose from to speed along test development.

**Comment by sriv** [comment]
#2433 raised for this.
